### PR TITLE
Add xdg icon theme support to the qt-gui on linux

### LINKED ISSUE #720: Support for xdg icon themes in qt-gui
So i looked into getting xdg icon theme support into the qt-gui and would like to hear the opinion of the rest of the team in regard of mapping icons that are not straight mappable to the [KDE](https://techbase.kde.org/Projects/Oxygen/namingSpec)/[GNOME](https://developer.gnome.org/icon-naming-spec/)/[XDG](https://specifications.freedesktop.org/icon-naming-spec/icon-naming-spec-latest.html) icon theme specifications.

Getting xdg icon theme support into the qt-gui allows people with impaired vision to use a icon set that works better for them. e.g. HighContrast Icon set

The current approach would look first if a theme icon is available, if not it would fallback to the icon shipped with the application.

fallback Icon | icon for | proposed icon name | Breeze | Oxygen Icon | Adwaita Icon | HC Icon | notes
-----------------|------------|-----------------|-------------------|-------------------|----------------|---------|-----
document-close | deleting key/array | edit-delete | edit-delete | edit-delete | edit-delete | edit-delete | original icon document-close is kde specific, edit-delete is part of xdg spec
|| _alternative_ | user-trash | user-trash | user-trash | user-trash | user-trash | 
|| _alternative_ | edit-clear | edit-clear | edit-clear | edit-clear | edit-clear
import | Import Configuration | document-open | document-open | document-open | ![document-open] (https://cloud.githubusercontent.com/assets/1131371/15274093/6881679c-1aab-11e6-8d61-a3338e013da9.png) | document-open | original icon document-import kde specific, document-open is part of xdg spec
export | Export Configuration | document-save | document-save | document-save | document-save | document-save | original icon document-export is kde specific, document-save is part of xdg spec
|| _alternative_ | document-save-as | document-save-as | document-save-as | document-save-as | document-save-as | maybe save-as is a better representation for export then save? 
mount | Mount Backend | list-add | list-add | list-add | list-add | list-add | This is the best i could find in the xdg spec for mount and unmount, they should be clear as long as they are shown in context.
unmount | Unmount Backend | list-remove | list-remove | list-remove | list-remove | list-remove
new-array | New Array | view-grid | view-grid | view-grid | | view-grid | Grid as abstract visualisation of array. Is this to much of a stretch?
color | Color Settings | preferences-desktop-color | preferences-desktop-color | preferences-desktop-color ||| kde specific
|| _alternative_ | preferences-color ||| preferences-color | preferences-color | gnome specific. Maybe find a solution to give the image provider more then one icon name
|| _alternative_ | applications-system | applications-system | applications-system | applications-system | preferences-system
edit-rename | edit key/array | edit-rename | edit-rename | edit-rename ||| kde specific
|| _alternative_ | accessories-text-editor | accessories-text-editor | accessories-text-editor | accessories-text-editor | accessories-text-editor
|| _alternative_ | document-properties | document-properties | document-properties | document-properties | document-properties
whats_this | contextual help | help-contextual | help-contextual | help-contextual | help-contents |  help-contents | help-contextual is kde specific it automaticly falls back to default help icon on gnome

### PR DESCRIPTION
**Pull request by sirblackheart**
Closes #720

This commit adds xdg icon support linux. This option is on by default
but can be toggled in the new Appearance Settings menu which has been
merged with the old ui color settings menu.

To have full xdg icon theme support a QQuickImageProvider has been
added. In addition the qt gui now depends on the qt5-svg module, as many
modern icon themes now provide svg.

Translation have been updated in regards of the changes.

Some typos in the desktop file have been fixed on the way.

### CODE REVIEW

**Comment by markus2330** [comment]
Looks very good for me.

A major change not discussed is: Why renaming the binary? Why installing it to bin?

Would be good to have separate commits for functional changes and typo fixes. You can leave it for now, but please do more commits the next time.

You can also add yourself in the "About Elektra Editor". Why not using the "Qt" name there?

@0003088 Can you also take a look if you like the changes? What do you think about the resizing the settings dialog?

File: src/tools/qt-gui/org.libelektra.elektra-qt-editor.desktop [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[Desktop Entry]
+Name=Elektra Qt Editor
+GenericName=Qt Configuration Editor for Elektra
```
Comment (markus2330) [inline src/tools/qt-gui/org.libelektra.elektra-qt-editor.desktop]: a bit confusing: does it allow to configure qt?

File: src/tools/qt-gui/qml/AppearanceSettingsWindow.qml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+						treeView.treeModel.refresh()
+					}
+				}
+			}
+		}
+	}
+	cancelButton.action.text: qsTr("&Close")
+	cancelButton.action.onTriggered: {
+		if (colorEdited)
+			guiSettings.setKDB()
```
Comment (markus2330) [inline src/tools/qt-gui/qml/AppearanceSettingsWindow.qml]: seems to be functional, should this be placed in the qml?

File: src/tools/qt-gui/qml/GUIActions.qml [hunk 3]
```diff
@@ -360,13 +376,15 @@ Item {
...
 		iconSource: "icons/help-about.png"
+		iconName: (guiSettings.useSystemIconTheme) ? "help-about" : ""
 		onTriggered: pluginInfo.show()
 	}
 
 	Action {
 		id: quitAction
 		text: qsTr("Quit")
 		iconSource: "icons/application-exit.png"
+		iconName: (guiSettings.useSystemIconTheme) ? "application-exit" : ""
```
Comment (markus2330) [inline src/tools/qt-gui/qml/GUIActions.qml]: what about writing a little function for this? Doesn't `useIcon("application-exit")` look nicer?

File: src/tools/qt-gui/qml/i18n/lang_de_DE.ts [hunk 4]
```diff
@@ -1,272 +1,552 @@
...
-<!DOCTYPE TS>
-<TS version="2.0" language="de_DE">
+< ?
+	xml version = "1.0" encoding = "utf-8" ?
+	> <!DOCTYPE TS><TS version = "2.1" language = "de_DE"><context><name>AboutWindow<
+		  /name>
+    <message>
+        <source>About Elektra Editor</source><translation type = "vanished">
+				Über Elektra Editor</translation>
+    </message><message><location filename = "../AboutWindow.qml" line =
```
Comment (markus2330) [inline src/tools/qt-gui/qml/i18n/lang_de_DE.ts]: It looks like the formatter did a poor job here.

File: src/tools/qt-gui/src/guisettings.cpp [hunk 5]
```diff
@@ -39,14 +39,16 @@ void GUISettings::setDefaults ()
...
+	m_useSystemIconTheme = true;
 }
 
 GUISettings::GUISettings (QObject * parentGUISettings)
 : QObject (parentGUISettings), m_profile ("/current/"), m_base ("/sw/elektra/qtgui/#0/"), m_highlightColorString ("color/highlight"),
   m_frameColorString ("color/frame"), m_nodeWKeyColorString ("color/node/with"), m_nodeWOKeyColorString ("color/node/without"),
-  m_legacyBase ("/sw/libelektra.org/qt-gui/#0/"), m_legacyHighlightColorString ("highlight_color"),
-  m_legacyFrameColorString ("frame_color"), m_legacyNodeWKeyColorString ("node_with_key_color"),
-  m_legacyNodeWOKeyColorString ("node_without_key_color")
+  m_useSystemIconThemeString ("icons/system"), m_legacyBase ("/sw/libelektra.org/qt-gui/#0/"),
```
Comment (markus2330) [inline src/tools/qt-gui/src/guisettings.cpp]: Singular (`icon/system`) seems to fit better with others.

File: src/tools/qt-gui/src/guisettings.cpp [hunk 6]
```diff
@@ -113,18 +121,34 @@ void GUISettings::setNodeWithoutKeyColor (const QColor & color)
...
+void GUISettings::appendColor (const std::string & keyName, const QColor & color)
 {
 	std::string name = "user" + m_base + m_profile + keyName;
 	m_config.append (Key (name, KEY_VALUE, color.name ().toStdString ().c_str (), KEY_END));
 }
 
+void GUISettings::appendBool (const std::string & keyName, const bool value)
+{
+	std::string name = "user" + m_base + m_profile + keyName;
+	m_config.append (Key (name, KEY_VALUE, (value ? "true" : "false"), KEY_END));
```
Comment (markus2330) [inline src/tools/qt-gui/src/guisettings.cpp]: Non-POD types (such as string) and vaargs are a bit dangerous.

What about using `key.set<bool>(value)`?

File: src/tools/qt-gui/src/guisettings.cpp [hunk 7]
```diff
@@ -141,6 +165,22 @@ void GUISettings::lookupColor (const std::string & keyName, QColor & toSet) cons
...
 
+void GUISettings::lookupBool (const std::string & keyName, bool & value) const
+{
+	Key key = m_config.lookup (keyName);
+
+	if (!key) return; // nothing to do
+
+	try
+	{
+		value = (key.get<std::string> ().compare ("true") == 0 ? true : false);
```
Comment (markus2330) [inline src/tools/qt-gui/src/guisettings.cpp]: Again, what about `.get<bool>`?

Unfortunately #308 is not implemented yet (a plugin that should bring different ways of saying true/false such as on/off to a canonical form)

File: src/tools/qt-gui/org.libelektra.elektra-qt-editor.desktop (hunk 1 above)
Comment (sirblackheart) [inline src/tools/qt-gui/org.libelektra.elektra-qt-editor.desktop]: Ok going to drop the Qt. It probably should not be in the generic name anyway.  The line is based on the one from dconf-editor desktop file.

File: src/tools/qt-gui/qml/AppearanceSettingsWindow.qml (hunk 2 above)
Comment (sirblackheart) [inline src/tools/qt-gui/qml/AppearanceSettingsWindow.qml]: Not sure, but those are the functions exposed to qml to actually do those things. And they are used for them in other places. But we probably should move 
```
guiSettings.setKDB()
treeView.treeModel.synchronize()
treeView.treeModel.refresh()
```
down into the close button trigger, so the tree view gets refreshed if we change any options. The rest of the main windows is inactive till we hit the close button so it should be fine to do it then.

File: src/tools/qt-gui/qml/GUIActions.qml (hunk 3 above)
Comment (sirblackheart) [inline src/tools/qt-gui/qml/GUIActions.qml]: Sure. I will look into how to write functions in qml.

File: src/tools/qt-gui/src/guisettings.cpp (hunk 6 above)
Comment (sirblackheart) [inline src/tools/qt-gui/src/guisettings.cpp]: I run into problems with that but will give it another try

File: src/tools/qt-gui/org.libelektra.elektra-qt-editor.desktop (hunk 1 above)
Comment (markus2330) [inline src/tools/qt-gui/org.libelektra.elektra-qt-editor.desktop]: Yes, according to [Arch Linux Wiki](https://wiki.archlinux.org/index.php/Desktop_entries) it should be "what you would generally call an application that does what this specific application offers", so rather "Elektra Configuration Editor" or even "Configuration Editor".

**Comment by sirblackheart** [comment]
> A major change not discussed is: Why renaming the binary? Why installing it to bin?

`TryExec` in the desktop entry did not work with `kdb qt-gui`. The `TryExec` is the entry that hides the application from start menus if it can not execute the `TryExec` parameter. We could just drop the `TryExec` to solve this.
The other reason to switch to a more clear bin name was to have a speaking name in process lists as `qt-gui` is not very clear for that.
But i just realized i also changed the bin name in the tool exec folder. That was not my intention and i will look into the possibility to have both binary with different names.

> Would be good to have separate commits for functional changes and typo fixes. You can leave it for now, but please do more commits the next time.

No you are right about that one. I will look into getting this into more structured commits, as git even gets problems with renames when to many changes happen.

File: src/tools/qt-gui/qml/GUIActions.qml (hunk 3 above)
Comment (markus2330) [inline src/tools/qt-gui/qml/GUIActions.qml]: Do not invest too much time here.

File: src/tools/qt-gui/src/guisettings.cpp (hunk 6 above)
Comment (markus2330) [inline src/tools/qt-gui/src/guisettings.cpp]: Again: Do not invest too much time here. Only make sure to fix the vaargs issue. (I think its undefined behaviour)

File: src/tools/qt-gui/qml/GUIActions.qml (hunk 3 above)
Comment (0003088) [inline src/tools/qt-gui/qml/GUIActions.qml]: qml uses ordinary javascript functions.

**Pull request closed without merging**